alock.1
.TH ALOCK 1
.SH NAME
alock \- Lock X display until password supplied
.SH SYNOPSIS
.B alock
.SH DESCRIPTION
.B alock
locks the X server till the user enters their password at the keyboard.

While
.B alock
is running, the mouse and keyboard are grabbed and the mouse cursor
becomes a padlock.  Output displayed by X programs, and windows put up
by new X clients, continue to be visible, and any new output is
displayed normally, except you run in -blank mode. Then the content
of the Screen is hidden behind a black Window.

The mouse and keyboard are returned when the user types their
password, followed by Enter or Newline.  If an incorrect password is
entered the bell is sounded.  Pressing Backspace or Delete erases one
character of a password partially typed; pressing Escape or Clear
clears anything that has been entered.

If too many attempts are made in too short a time further keystrokes
generate bells and are otherwise ignored until a timeout has expired.

The X server screen saver continues to operate normally; if it comes
into operation the display may be restored by the usual means of
touching a key (Shift, for example) or the mouse.
.SH OPTIONS, X RESOURCES, CONFIGURATION
None.
.SH BUGS
Additional input devices other than the keyboard and mouse are not
disabled.

The timeouts, bitmaps and mouse cursor colour cannot be modifed.
.SH SEE ALSO
.BR X "(1), " xlock "(1), Xlib Documentation."
.SH AUTHOR
Mathias Gumz <akira at gmail dot com>
Ian Jackson <ijackson@nyx.cs.du.edu>
.SH COPYRIGHT
Copyright (C) 2005 Mathias Gumz
Copyright (C) 1993, 1994 by Ian Jackson.

Permission to use, copy, modify and distribute this software and its
documentation is granted under the terms of the GNU General Public
Licence, version 2 or later, as published by the Free Software
Foundation.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.


CHANGELOG
Version 0.5

2005-05-14:

	* started documention, using asciidoc

2005-05-09:

	* changed project-name to 'alock':
		+ simpler, better name
		+ i want to dedicate it to my gf: anja
	* minor fixes

2005-05-08:

	* changed the auth_md5.c - code to the openbsd-version (public domain)
	* added auth_sha1.c - also from openbsd-codebase, also public domain
	* added two stand-alone programs to calculate md5/sha1 hashs:
		+ src/asha1 - reads from stdin, EOF ends
		+ src/amd5 - reads from stdin, EOF ends
		(just in case your system DOESNT have md5/md5sum or sha1/sha1sum
	* minor changes to aklock.c

2005-05-07:

	* big restructering + feature enhancements
		+ added auth_<method>.c, each basic auth-method has now
		  its own implementation-file
		+ added -auth - flag:
			- list      - lists all available methods
			- none      - no auth, like -nolock
			- passwd    - standard login, possible shadowsupport
			- pam       - use pam for authentification
			- md5:hash  - provide a md5-hash for authentification
	* support for 'shadow'-passwords is part of the passwd authmodule

2005-05-06:

	* added -nolock
	* discard all pending xevents after wrong password to prevent
	  strange folks for blocking the current session by filling
	  the eventqueue with wrong authentification and thus block
	  the real authentification (kind of denial of service :)
	* added suuport to load / have different cursor-shapes
		+ -cursor theme|<format>:<filename>
	* added support for xcursor
	* restructuring
		+ contrib/ -> for stuff coming from the "outside"
		  contrib/xcursor-watch -> xcursor taken from 5532-BlueGlass-XCursors-3D-0.4.tar.bz2
		+ src/

2005-05-02:

	* restructuring

2005-04-30:

	* added -cursor <theme>
	* cursor-bitmap-themes are stored in bitmaps/
	* started svn-usage

Version 0.1

2005-04-28: 
	* first release (0.1)
	* added pam-support (works ok with linux + freebsd so far)
	* use scons for building the stuff

# vim:ft=changelog


src/alock.c
/*------------------------------------------------------------------*\

    file: alock.c

        X Transparent Lock

    copyright:

        Copyright (C)2005 Mathias Gumz (forked alock)
        Copyright (C)1993,1994 Ian Jackson   (xtrlock)

    license:

        This is free software; you can redistribute it and/or modify
        it under the terms of the GNU General Public License as published by
        the Free Software Foundation; either version 2, or (at your option)
        any later version.

        This is distributed in the hope that it will be useful,
        but WITHOUT ANY WARRANTY; without even the implied warranty of
        MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
        GNU General Public License for more details.

\*------------------------------------------------------------------*/

/*------------------------------------------------------------------*\
\*------------------------------------------------------------------*/
#include <X11/X.h>
#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/keysym.h>
#include <X11/Xos.h>
#include <X11/cursorfont.h>

#ifdef HAVE_XPM
#   include <X11/xpm.h>
#endif /* HAVE_XPM */
#ifdef HAVE_XCURSOR
#   include <X11/Xcursor/Xcursor.h>
#endif /* HAVE_XCURSOR */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>

/*----------------------------------------------*\
\*----------------------------------------------*/

#include "alock.h"

struct aXInfo {

    Display* display;
    Window   root;
    Window   window;
    Colormap colormap;

    Cursor   cursor;

    int width;
    int height;
};
/*------------------------------------------------------------------*\
    globals
\*------------------------------------------------------------------*/

#define TIMEOUTPERATTEMPT 30000
#define MAXGOODWILL  (TIMEOUTPERATTEMPT*5)
#define INITIALGOODWILL MAXGOODWILL
#define GOODWILLPORTION 0.3

#ifdef DEBUG
#    define DBGMSG fprintf(stderr, "%s : %d\n", __FUNCTION__, __LINE__); fflush(stderr)
#else
#    define DBGMSG
#endif // DEBUG

static struct aAuth* alock_authmodules[] = {
#ifdef PAM_PWD
    &alock_auth_pam,
#endif
#ifdef PASSWD_PWD
    &alock_auth_passwd,
#endif /* PASSWD_PWD */
#ifdef HASH_PWD
    &alock_auth_md5,
    &alock_auth_sha1,
#endif /* HASH_PWD */
    &alock_auth_none,
    NULL
};

/*------------------------------------------------------------------*\
\*------------------------------------------------------------------*/

void displayUsage() {
    printf("\n");
    printf("alock [-h] [-v] [-blank] [-cursor <theme|xcursor:file>]\n");
    printf("       [-auth list|<none");
#ifdef PASSWD_PWD
    printf("|passwd");
#endif /* PASSWD_PWD */
#ifdef PAM_PWD
    printf("|pam");
#endif /* PAM_PWD */
#ifdef HASH_PWD
    printf("|md5:pwdhash|sha1:pwdhash");
#endif /* HASH_PWD */
    printf(">]\n");
}

/*------------------------------------------------------------------*\
\*------------------------------------------------------------------*/

void initOpts(struct aOpts* opts) {

    opts->auth = alock_authmodules[0];
    opts->use_blank = 0;

    opts->cursor_name = "mini";

    opts->color_bg = "steelblue3";
    opts->color_fg = "grey25";

}

void initXInfo(struct aXInfo* xinfo, struct aOpts* opts) {


    Display* dpy = XOpenDisplay(NULL);
    XColor color_fg;
    XColor color_bg;
    XColor tmp_color;
    Pixmap pixmap_cursor;
    Pixmap pixmap_cursor_mask;
    XWindowAttributes xgwa;
    struct aCursor* cursor = NULL;

    if (!dpy) {
        perror("alock: error, can't open connection to X");
        exit(1);
    }

    xinfo->display = dpy;
    xinfo->window = 0;

    xinfo->root = DefaultRootWindow(dpy);
    xinfo->colormap = DefaultColormap(dpy, DefaultScreen(dpy));

    XGetWindowAttributes(dpy, xinfo->root, &xgwa);

    if (opts->use_blank) {
        xinfo->width = xgwa.width;
        xinfo->height = xgwa.height;
    } else {
        xinfo->width = 1;
        xinfo->height = 1;
    }

    if((XAllocNamedColor(dpy, xinfo->colormap, opts->color_bg, &tmp_color, &color_bg)) == 0)
        XAllocNamedColor(dpy, xinfo->colormap, "black", &tmp_color, &color_bg);
    if((XAllocNamedColor(dpy, xinfo->colormap, opts->color_fg, &tmp_color, &color_fg)) == 0)
        XAllocNamedColor(dpy, xinfo->colormap, "white", &tmp_color, &color_fg);

    /* load cursors */
#ifdef HAVE_XCURSOR
    if (opts->cursor_name && (strstr(opts->cursor_name, "xcursor:"))) {
        Cursor xcursor;
        if ((xcursor = XcursorFilenameLoadCursor(dpy, &opts->cursor_name[8]))) {
            xinfo->cursor = xcursor;
            return;
        } else {
            printf("alock: error, couldnt load [%s]\n", &opts->cursor_name[8]);
        }
    }
#endif /* HAVE_XCURSOR */
    /* create cursor from X11/cursorfont.h */
    if (opts->cursor_name && (strstr(opts->cursor_name, "font:"))) {
        Cursor fcursor;
        if ((fcursor = XCreateFontCursor(dpy, atoi(&opts->cursor_name[5])))) {
            XRecolorCursor(dpy, fcursor, &color_fg, &color_bg);
            xinfo->cursor = fcursor;
            return;
        } else
            printf("alock: error, couldnt creat fontcursor [%d].\n", atoi(&opts->cursor_name[5]));

    }
    /* TODO: doesnt work yet. */
/*
 * if (opts->cursor_name && (strstr(opts->cursor_name, "xbm:"))) {
        unsigned int w, h, xhot, yhot;
        if (XReadBitmapFile(dpy, xinfo->root, &opts->cursor_name[4],
                            &w, &h, &pixmap_cursor, &xhot, &yhot)) {
            xinfo->cursor = XCreatePixmapCursor(dpy,
                                                pixmap_cursor, NULL,
                                                &color_fg, &color_bg,
                                                xhot, yhot);
            return;
        } else {
            printf("alock: error, couldnt load [%s]\n", &opts->cursor_name[4]);
        }
    }
   */
    /* look internal cursors */
    for(cursor = alock_cursors; cursor->name != NULL; cursor++) {
        if (!strcmp(cursor->name, opts->cursor_name))
                break;
    }

    if (!cursor->name)
        cursor = alock_cursors;

    pixmap_cursor = XCreateBitmapFromData(dpy, xinfo->root, cursor->bits, cursor->width, cursor->height);
    pixmap_cursor_mask = XCreateBitmapFromData(dpy, xinfo->root, cursor->mask, cursor->width, cursor->height);

    xinfo->cursor = XCreatePixmapCursor(dpy,
                                        pixmap_cursor, pixmap_cursor_mask,
                                        &color_fg, &color_bg,
                                        cursor->x_hot, cursor->y_hot);


}

int main(int argc, char **argv) {

    XEvent ev;
    KeySym ks;
    char cbuf[10], rbuf[50];
    int clen, rlen = 0;
    long goodwill = INITIALGOODWILL, timeout = 0;

    XSetWindowAttributes xswa;
    long xsmask = 0;

    struct aXInfo xinfo;
    struct aOpts opts;

    int arg = 0;

    initOpts(&opts);

    /*  parse options */
    if (argc != 1) {
        for(arg = 1; arg <= argc; arg++) {
            if (!strcmp(argv[arg - 1], "-blank")) {
                opts.use_blank = 1;
            } else if (!strcmp(argv[arg - 1], "-auth")) {
                if (arg < argc) {

                    char* char_tmp;
                    struct aAuth* auth_tmp = NULL;
                    struct aAuth** i;
                    if (!strcmp(argv[arg], "list")) {
                        for(i = alock_authmodules; *i; ++i) {
                            printf("%s\n", (*i)->name);
                        }
                        exit(0);
                    }

                    for(i = alock_authmodules; *i; ++i) {
                        char_tmp = strstr(argv[arg], (*i)->name);
                        if(char_tmp && char_tmp == argv[arg]) {
                            auth_tmp = (*i);
                            if (!auth_tmp->init(argv[arg])) {
                                fprintf(stderr, "alock: error, failed init of [%s].\n", auth_tmp->name);
                                exit(1);
                            }
                            opts.auth = auth_tmp;
                            ++arg;
                            break;
                        }
                    }

                    if (!auth_tmp) {
                        fprintf(stderr, "alock: error, couldnt find the auth module you specified.\n");
                        exit(1);
                    }

                } else {
                    fprintf(stderr, "alock, error, missing argument\n");
                    displayUsage();
                    exit(1);
                }
            } else if (!strcmp(argv[arg - 1], "-cursor")) {
                if (arg < argc)
                    opts.cursor_name = argv[arg];
                else {
                    printf("alock: error, missing argument\n");
                    displayUsage();
                    exit(1);
                }
            } else if (!strcmp(argv[arg - 1], "-h")) {
                displayUsage();
                exit(0);
            } else if (!strcmp(argv[arg - 1], "-v")) {
                printf("alock-%s by m.gumz 2005\n", VERSION);
                exit(0);
            }
        }
    }

    initXInfo(&xinfo, &opts);

    /* create the windows */
    xswa.override_redirect = True;
    xsmask |= CWOverrideRedirect;

    if (opts.use_blank) {

        xswa.background_pixel = BlackPixel(xinfo.display, DefaultScreen(xinfo.display));
        xswa.colormap = xinfo.colormap;
        xsmask |= CWBackPixel;
        xsmask |= CWColormap;
    }

    xinfo.window = XCreateWindow(xinfo.display, xinfo.root,
                          0, 0, xinfo.width, xinfo.height,
                          0, /* borderwidth */
                          CopyFromParent, /* depth */
                          InputOutput, /* class */
                          CopyFromParent, /* visual */
                          xsmask, &xswa);



    XSelectInput(xinfo.display, xinfo.window, KeyPressMask|KeyReleaseMask);
    XMapWindow(xinfo.display, xinfo.window);
    XRaiseWindow(xinfo.display, xinfo.window);
    
    /* TODO: -bg <blank|transparent|shaded> */
 /*   if (opts.use_blank) {
        XImage* ximage;
                
        ximage = XGetImage (xinfo.display, xinfo.root, 0, 0,
                    xinfo.width, xinfo.height, AllPlanes, ZPixmap);

        if (ximage) {
            GC gc;
            XGCValues xgcv;
            xgcv.background = BlackPixel(xinfo.display, DefaultScreen(xinfo.display));
            xgcv.foreground = 0;
            
            gc = XCreateGC(xinfo.display, xinfo.window, GCForeground|GCBackground, &xgcv);

            XPutImage(xinfo.display, xinfo.window, 
                      gc, 
                      ximage, 
                      0, 0, 
                      0, 0, xinfo.width, xinfo.height);
            XDestroyImage(ximage);
            XFreeGC(xinfo.display, gc);
        }
    } */
    /* try to grab 2 times, another process (windowmanager) may have grabbed
     * the keyboard already */
    if ((XGrabKeyboard(xinfo.display, xinfo.window, True, GrabModeAsync, GrabModeAsync,
                      CurrentTime)) != GrabSuccess) {
        sleep(1);
        if ((XGrabKeyboard(xinfo.display, xinfo.window, True, GrabModeAsync, GrabModeAsync,
                        CurrentTime)) != GrabSuccess) {
            perror("alock: couldnt grab the keyboard.\n");
            exit(1);
        }
    }

    if (XGrabPointer(xinfo.display, xinfo.window, False, (KeyPressMask|KeyReleaseMask) & 0,
                     GrabModeAsync, GrabModeAsync, None, xinfo.cursor, CurrentTime) != GrabSuccess) {
        XUngrabKeyboard(xinfo.display, CurrentTime);
        perror("alock: couldnt grab the pointer.\n");
        exit(1);
    }

    /* eventhandling */
    for(;;) {
        XNextEvent(xinfo.display, &ev);
        switch (ev.type) {
        case KeyPress:
            
            if (ev.xkey.time < timeout) {
                XBell(xinfo.display, 0);
                break;
            }

            clen = XLookupString(&ev.xkey, cbuf, 9, &ks, 0);
            switch (ks) {
            case XK_Escape:
            case XK_Clear:
                rlen = 0;
                break;
            case XK_Delete:
            case XK_BackSpace:
                if (rlen > 0)
                    rlen--;
                break;
            case XK_Linefeed:
            case XK_Return:
                if (rlen == 0)
                    break;
                rbuf[rlen] = 0;

                if (opts.auth->auth(rbuf))
                    goto exit;
                
                XSync(xinfo.display, True); /* discard pending events to start really fresh */
                XBell(xinfo.display, 0);
                rlen = 0;
                if (timeout) {
                    goodwill += ev.xkey.time - timeout;
                    if (goodwill > MAXGOODWILL) {
                        goodwill = MAXGOODWILL;
                    }
                }
                timeout = -goodwill * GOODWILLPORTION;
                goodwill += timeout;
                timeout += ev.xkey.time + TIMEOUTPERATTEMPT;
                break;
            default:
                if (clen != 1)
                    break;
                if (rlen < sizeof(rbuf))
                    rbuf[rlen] = cbuf[0];
                rlen++;
                break;
            }
            break;
        default:
                break;
        }
    }

exit:

    opts.auth->deinit();

    XDestroyWindow(xinfo.display, xinfo.window);
    XFreeCursor(xinfo.display, xinfo.cursor);
    XCloseDisplay(xinfo.display);

    return 0;
}

